buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java

```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
 
     private void logFileContents(String description, Path from) {
         logger.error("{} `{}`", description, this);
+        long lineCutoff;
         try(Stream<String> lines = Files.lines(from, StandardCharsets.UTF_8)) {
-            lines
-                .map(line -> "  " + line)
-                .forEach(logger::error);
+            lineCutoff = lines.count();
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        final Map<String, Long> errorsAndWarnings;
+        try(Stream<String> lines = Files.lines(from, StandardCharsets.UTF_8)) {
+            errorsAndWarnings = lines
+                .filter(line -> line.contains("ERROR") || line.contains("WARN"))
+                .map(this::normalizeLogLine)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        lineCutoff -= 50;
+        try (LineNumberReader reader = new LineNumberReader(Files.newBufferedReader(from))) {
+            String line = null;
+            do {
+                line = reader.readLine();
+                if (line == null) {
+                    continue;
+                }
+                String normalizedLine = normalizeLogLine(line);
+                if (errorsAndWarnings.keySet().contains(normalizedLine)) {
+                    logger.lifecycle("  " + line);
+                    if (errorsAndWarnings.get(normalizedLine) != 1) {
+                        logger.lifecycle("    ↑ repeated another " + errorsAndWarnings.get(normalizedLine) + " times ↑");
+                    }
+                    errorsAndWarnings.remove(normalizedLine);
+                }
+                if (reader.getLineNumber() == lineCutoff) {
```

Comment:
Why have a cutoff? If there are errors, they are possibly relevant, and more likely more relevant closer to the end of the log (or possibly the first error). I think we should continue to output all errors/warnings.

Reply:
We do not cut off errors or warnings. We show the last 50 lines of the log regardless of what it contains and show any error and warning before that.